src/keyboard/mod.rs

```diff
@@ -569,59 +612,252 @@ where
+                                        }
+                                        let elapsed_time = time_tracker.elapsed();
+                                        thread_repeat_impl.lock().unwrap().receive(
+                                            KeyRepeatEvent {
+                                                time: time
+                                                    + elapsed_time.as_secs() as u32 * 1000
+                                                    + elapsed_time.subsec_nanos() / 1000000,
+                                                modifiers: thread_modifiers,
+                                                rawkey: key,
+                                                keysym: thread_sym,
+                                                utf8: thread_utf8.clone(),
+                                            },
+                                            proxy.clone(),
+                                        );
+                                        // Rate
+                                        thread::sleep(Duration::from_millis(repeat_timing.0));
+                                        match thread_kill_chan.lock().unwrap().1.try_recv() {
+                                            Ok(_) | Err(mpsc::TryRecvError::Disconnected) => {
+                                                break
+                                            }
+                                            _ => {}
+                                        }
+                                    }
+                                });
+                            }
+                        }
+                    } else {
+                        if key_held == Some(key) {
+                            kill_chan.lock().unwrap().0.send(()).unwrap();
+                            key_held = None;
+                        }
+                        event_impl.receive(
+                            Event::Key {
+                                serial,
+                                time,
+                                modifiers,
+                                rawkey: key,
+                                keysym: sym,
+                                state: key_state,
+                                utf8: utf8.clone(),
+                            },
+                            proxy.clone(),
+                        );
+                    }
                 }
                 wl_keyboard::Event::Modifiers {
                     mods_depressed,
                     mods_latched,
                     mods_locked,
                     group,
                     ..
-                } => state.update_modifiers(mods_depressed, mods_latched, mods_locked, group),
+                } => {
+                    state.update_modifiers(mods_depressed, mods_latched, mods_locked, group);
+                    if key_held.is_some() {
+                        state_chan.lock().unwrap().0.send(()).unwrap();
+                    }
```

Comment:
This code still means that pressing a modifier will stop any repeat, right? I don't think that's the expected behavior.

If I hold `a` on my computer and press then release `<shift>` several times, I get something like `aaaaaaaaaaaaaAAAAAAAAAAaaaaaaaaaaAAAAAAaaaaaaaaaAAAAAaaaaaaaa`.

Reply:
Maybe you are thinking of **kill**_chan instead of **state**_chan, state_chan simply requests the thread to update modifiers when they are changed rather then updating the values every 30ms. I don't really know whether this is much more performant then just updating them for every key repetition though.

Reply:
Actually re-reading that, is 
> aaaaaaaaaaaaaAAAAAAAAAAaaaaaaaaaaAAAAAAaaaaaaaaaAAAAAaaaaaaaa

not the expected behavior? what would be?

Reply:
> Maybe you are thinking of kill_chan instead of state_chan

Ah right, my bad. I've read it too quickly. No problem then.